karma/module.py

```diff
@@ -0,0 +1,641 @@
+        reaction_author: Tuple[int, int] = (reaction.guild_id, reaction.member.id)
+
+        return (message_author, reaction_author, emoji_value)
+
+    @commands.Cog.listener()
+    async def on_raw_reaction_add(self, reaction: discord.RawReactionActionEvent):
+        """Handle added reactions."""
+        author_m, author_r, emoji_value = await self.karma_cache_check(reaction)
+
+        self.value_cache.setdefault(author_m, 0)
+        self.value_cache[author_m] += emoji_value
+
+        if emoji_value > 0:
+            self.given_cache.setdefault(author_r, 0)
+            self.given_cache[author_r] += emoji_value
+        else:
+            self.taken_cache.setdefault(author_r, 0)
+            self.taken_cache[author_r] += -emoji_value
+
+    @commands.Cog.listener()
+    async def on_raw_reaction_remove(self, reaction: discord.RawReactionActionEvent):
+        """Handle removed reactions."""
+        author_m, author_r, emoji_value = await self.karma_cache_check(reaction)
+
+        self.value_cache.setdefault(author_m, 0)
+        self.value_cache[author_m] -= emoji_value
+
+        if emoji_value > 0:
+            self.given_cache.setdefault(author_r, 0)
+            self.given_cache[author_r] -= emoji_value
+        else:
+            self.taken_cache.setdefault(author_r, 0)
+            self.taken_cache[author_r] -= -emoji_value
+
+    #
+
+    @commands.check(check.acl)
+    @commands.group(name="karma")
+    async def karma_(self, ctx):
+        await utils.Discord.send_help(ctx)
+
+    @commands.check(check.acl)
+    @karma_.command(name="get")
+    async def karma_get(self, ctx, member: Optional[discord.Member] = None):
+        """Display karma information on some user."""
+        if member is None:
+            member = ctx.author
+        kmember = KarmaMember.get_or_add(ctx.guild.id, member.id)
+
+        embed = utils.Discord.create_embed(
+            author=ctx.author,
+            title=_(ctx, "User karma"),
+            description=utils.Text.sanitise(member.display_name),
+        )
+
+        embed.add_field(
+            name=_(ctx, "Total value"),
+            value=f"**{kmember.value}** (#{kmember.value_position})",
+            inline=False,
+        )
+        embed.add_field(
+            name=_(ctx, "Karma given"),
+            value=f"**{kmember.given}** (#{kmember.given_position})",
+        )
+        embed.add_field(
+            name=_(ctx, "Karma taken"),
+            value=f"**{kmember.taken}** (#{kmember.taken_position})",
+        )
+
+        avatar_url: str = member.display_avatar.replace(size=256).url
+        embed.set_thumbnail(url=avatar_url)
+
+        await ctx.reply(embed=embed)
+
+    @commands.check(check.acl)
+    @karma_.command(name="emoji")
+    async def karma_emoji(self, ctx, emoji: str):
+        """Display karma information on emoji."""
+        with contextlib.suppress(commands.EmojiNotFound):
+            emoji = await commands.EmojiConverter().convert(ctx, emoji)
```

Comment:
Can we use `emoji.is_discord_emoji()` instead, assuming we change the type hint above to `discord.Emoji`?

Reply:
Do you mean is_custom_emoji?

https://discordpy.readthedocs.io/en/stable/api.html#discord.PartialEmoji.is_custom_emoji

Reply:
I think this depends on if we would be able to get PartialEmoji instead of Emoji. Maybe class `PartialEmojiConverter` would be useful.

Reply:
OK, I've found out that you would have to convert emoji to PartialEmoji, which could raise exception too. There could be a way through `Union[discord.Emoji, discord.PartialEmoji, str]`